domain/application/resource/store_test.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package resource
+
+import (
+    "bytes"
+    "context"
+    "crypto/sha512"
+    "fmt"
+    "io"
+
+    jc "github.com/juju/testing/checkers"
+    "go.uber.org/mock/gomock"
+    gc "gopkg.in/check.v1"
+
+    charmresource "github.com/juju/juju/internal/charm/resource"
+)
+
+type resourceStoreSuite struct {
+    objectStore            *MockObjectStore
+    modelObjectStoreGetter *MockModelObjectStoreGetter
+    resource               Resource
+    file                   io.ReadCloser
+}
+
+var _ = gc.Suite(&resourceStoreSuite{})
+
+func (s *resourceStoreSuite) SetUpTest(c *gc.C) {
+    data := "spamspamspam"
+    size := int64(len(data))
+    sha384hash := sha512.New384()
+    sha384hash.Write([]byte(data))
+    fp := fmt.Sprintf("%x", sha384hash.Sum(nil))
+    fingerprint, err := charmresource.ParseFingerprint(fp)
+    c.Assert(err, jc.ErrorIsNil)
+    s.resource = Resource{
+        Resource: charmresource.Resource{
+            Meta: charmresource.Meta{
+                Name: "spam-resource",
+                Type: charmresource.TypeFile,
+            },
+            Fingerprint: fingerprint,
+            Size:        size,
+        },
+        ApplicationID: "fake-app-uuid",
+    }
+    s.file = io.NopCloser(bytes.NewBufferString(data))
+}
+
+func (s *resourceStoreSuite) setupMocks(c *gc.C) *gomock.Controller {
+    ctrl := gomock.NewController(c)
+
+    s.modelObjectStoreGetter = NewMockModelObjectStoreGetter(ctrl)
+    s.objectStore = NewMockObjectStore(ctrl)
+
+    return ctrl
+}
+
+func (s *resourceStoreSuite) TestObjectStoreGetter(c *gc.C) {
+    defer s.setupMocks(c).Finish()
+
+    factory := NewResourceStoreFactory(s.modelObjectStoreGetter)
+    s.modelObjectStoreGetter.EXPECT().GetObjectStore(gomock.Any()).Return(s.objectStore, nil)
+
+    store, err := factory.GetResourceStore(context.Background(), s.resource.Resource.Meta.Type)
+    c.Assert(err, jc.ErrorIsNil)
+    c.Assert(store, gc.Equals, s.objectStore)
+}
+
+func (s *resourceStoreSuite) TestObjectStoreGetterAddStore(c *gc.C) {
+    defer s.setupMocks(c).Finish()
+
+    factory := NewResourceStoreFactory(s.modelObjectStoreGetter)
+    newStore := FileResourceStore{objectStore: s.objectStore}
+    factory.AddStore(charmresource.TypeContainerImage, newStore)
+
+    store, err := factory.GetResourceStore(context.Background(), charmresource.TypeContainerImage)
+    c.Assert(err, gc.IsNil)
+    c.Assert(store, gc.Equals, newStore)
+}
+
+func (s *resourceStoreSuite) TestObjectStoreGetterNotFound(c *gc.C) {
+    defer s.setupMocks(c).Finish()
+
+    factory := NewResourceStoreFactory(s.modelObjectStoreGetter)
+
+    _, err := factory.GetResourceStore(context.Background(), charmresource.Type(0))
+    c.Assert(err, gc.NotNil)
```

Comment:
**question:** No specific error check ?

Reply:
This is not a typed error so it would have to be string matching. We could do that, but it always feels a bit icky to me. Fair enough if you would rather I do that though.